Failed to compile benches: no method named `slice` found

Hi there 👋 

I am trying to run benchmarks and getting bunch of error regarding missing method names:
```
$ cargo bench
...
error[E0599]: no method named `slice` found for struct `Repeated` in the current scope
   --> benches/lex.rs:129:53
    |
129 |         let digits = one_of(b'0'..=b'9').repeated().slice();
    |                                                     ^^^^^ method not found in `Repeated<OneOf<RangeInclusive<u8>, _, _>, _, _, _>`
...
```

Is there a way to get it working? I haven't really looked at the code precisely trying to fix the error, I had hoped that someone with knowledge of the library can give some tips.

Thanks!
Oleg

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -23,11 +23,11 @@ jobs:
             toolchain: nightly
             components: rustfmt, clippy
       - name: Run cargo check (all features)
-        run: cargo check --tests --verbose --all-features
+        run: cargo check --benches --examples --tests --verbose --all-features
       - name: Run cargo check (no features)
-        run: cargo check --tests --verbose --no-default-features
+        run: cargo check --benches --examples --tests --verbose --no-default-features
       - name: Run cargo clippy
-        run: cargo clippy --verbose --all-features -- -D warnings
+        run: cargo clippy --benches --examples --tests --verbose --all-features -- -D warnings
       - name: Run cargo fmt
         run: cargo fmt --verbose --check
       - name: Run cargo doc

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -85,7 +85,7 @@ unicode-ident =  "1.0.10"
 ariadne = "0.2"
 pom = "3.2"
 nom = "7.1"
-winnow = "0.5.0"
+winnow = "0.5.19"
 serde_json = { version = "1.0", features = ["preserve_order"] }
 ciborium = { version = "0.2" }
 criterion = "0.4.0"

--- a/benches/cbor.rs
+++ b/benches/cbor.rs
@@ -3,7 +3,7 @@ use std::hint::black_box;
 
 mod utils;
 
-static CBOR: &'static [u8] = include_bytes!("samples/sample.cbor");
+static CBOR: &[u8] = include_bytes!("samples/sample.cbor");
 
 fn bench_cbor(c: &mut Criterion) {
     // c.bench_function("cbor_nom", {
@@ -125,7 +125,8 @@ mod chumsky_zero_copy {
                     };
                     cfg.exactly(num)
                 }))
-                .map_slice(int_out);
+                .to_slice()
+                .map(int_out);
 
             let uint = read_int.map(CborZero::Int);
             let nint = read_int.map(|i| CborZero::Int(-1 - i));
@@ -134,14 +135,16 @@ mod chumsky_zero_copy {
                 any()
                     .repeated()
                     .configure(|cfg, ctx| cfg.exactly(*ctx as usize))
-                    .map_slice(CborZero::Bytes),
+                    .to_slice()
+                    .map(CborZero::Bytes),
             );
 
             let str = read_int.ignore_with_ctx(
                 any()
                     .repeated()
                     .configure(|cfg, ctx| cfg.exactly(*ctx as usize))
-                    .map_slice(|slice| CborZero::String(std::str::from_utf8(slice).unwrap())),
+                    .to_slice()
+                    .map(|slice| CborZero::String(std::str::from_utf8(slice).unwrap())),
             );
 
             let array = read_int.ignore_with_ctx(

--- a/benches/json.rs
+++ b/benches/json.rs
@@ -1,3 +1,5 @@
+#![allow(clippy::result_large_err, clippy::type_complexity)]
+
 use criterion::{black_box, criterion_group, criterion_main, Criterion};
 
 mod utils;
@@ -22,7 +24,7 @@ pub enum JsonZero<'a> {
     Object(Vec<(&'a [u8], JsonZero<'a>)>),
 }
 
-static JSON: &'static [u8] = include_bytes!("samples/sample.json");
+static JSON: &[u8] = include_bytes!("samples/sample.json");
 
 fn bench_json(c: &mut Criterion) {
     c.bench_function("json_nom", {
@@ -142,7 +144,8 @@ mod chumsky_zero_copy {
                 .then(int)
                 .then(frac.or_not())
                 .then(exp.or_not())
-                .map_slice(|bytes| str::from_utf8(bytes).unwrap().parse().unwrap())
+                .to_slice()
+                .map(|bytes| str::from_utf8(bytes).unwrap().parse().unwrap())
                 .boxed();
 
             let escape = just(b'\\').then_ignore(one_of(b"\\/\"bfnrt"));
@@ -150,7 +153,7 @@ mod chumsky_zero_copy {
             let string = none_of(b"\\\"")
                 .or(escape)
                 .repeated()
-                .slice()
+                .to_slice()
                 .delimited_by(just(b'"'), just(b'"'))
                 .boxed();
 
@@ -197,14 +200,14 @@ mod pom {
     }
 
     fn number() -> Parser<u8, f64> {
-        let integer = one_of(b"123456789") - one_of(b"0123456789").repeat(0..) | sym(b'0');
+        let integer = (one_of(b"123456789") - one_of(b"0123456789").repeat(0..)) | sym(b'0');
         let frac = sym(b'.') + one_of(b"0123456789").repeat(1..);
         let exp = one_of(b"eE") + one_of(b"+-").opt() + one_of(b"0123456789").repeat(1..);
         let number = sym(b'-').opt() + integer + frac.opt() + exp.opt();
         number
             .collect()
             .convert(str::from_utf8)
-            .convert(|s| f64::from_str(&s))
+            .convert(f64::from_str)
     }
 
     fn string() -> Parser<u8, String> {
@@ -237,10 +240,10 @@ mod pom {
         (seq(b"null").map(|_| Json::Null)
             | seq(b"true").map(|_| Json::Bool(true))
             | seq(b"false").map(|_| Json::Bool(false))
-            | number().map(|num| Json::Num(num))
-            | string().map(|text| Json::Str(text))
-            | array().map(|arr| Json::Array(arr))
-            | object().map(|obj| Json::Object(obj)))
+            | number().map(Json::Num)
+            | string().map(Json::Str)
+            | array().map(Json::Array)
+            | object().map(Json::Object))
             - space()
     }
 
@@ -344,7 +347,7 @@ mod nom {
         terminated(value, space)(i)
     }
 
-    pub fn json<'a>(i: &'a [u8]) -> IResult<&'a [u8], JsonZero, (&'a [u8], nom::error::ErrorKind)> {
+    pub fn json(i: &[u8]) -> IResult<&[u8], JsonZero, (&[u8], nom::error::ErrorKind)> {
         root(i)
     }
 }
@@ -352,7 +355,7 @@ mod nom {
 mod winnow {
     use winnow::{
         ascii::{digit0, digit1, escaped},
-        combinator::separated0,
+        combinator::separated,
         combinator::{alt, dispatch},
         combinator::{cut_err, fail, opt, peek},
         combinator::{preceded, separated_pair, terminated},
@@ -403,7 +406,7 @@ mod winnow {
         preceded(
             '[',
             cut_err(terminated(
-                separated0(value, preceded(space, ',')),
+                separated(0.., value, preceded(space, ',')),
                 preceded(space, ']'),
             )),
         )
@@ -427,7 +430,7 @@ mod winnow {
         preceded(
             '{',
             cut_err(terminated(
-                separated0(member, preceded(space, ',')),
+                separated(0.., member, preceded(space, ',')),
                 preceded(space, '}'),
             )),
         )

--- a/benches/lex.rs
+++ b/benches/lex.rs
@@ -26,7 +26,7 @@ pub enum Token<'a> {
     Comma,
 }
 
-static SAMPLE: &'static [u8] = include_bytes!("tokens.txt");
+static SAMPLE: &[u8] = include_bytes!("tokens.txt");
 
 fn bench_lex(c: &mut Criterion) {
     c.bench_function("lex_chumsky_zero_copy", {
@@ -126,7 +126,7 @@ mod chumsky_zero_copy {
     use std::str;
 
     pub fn parser<'a>() -> impl Parser<'a, &'a [u8], Vec<Token<'a>>> {
-        let digits = one_of(b'0'..=b'9').repeated().slice();
+        let digits = one_of(b'0'..=b'9').repeated().to_slice();
 
         let int = one_of(b'1'..=b'9')
             .repeated()
@@ -148,7 +148,8 @@ mod chumsky_zero_copy {
             .then(int)
             .then(frac.or_not())
             .then(exp.or_not())
-            .map_slice(|bytes| str::from_utf8(bytes).unwrap().parse().unwrap())
+            .to_slice()
+            .map(|bytes| str::from_utf8(bytes).unwrap().parse().unwrap())
             .boxed();
 
         let escape = just(b'\\')
@@ -169,11 +170,11 @@ mod chumsky_zero_copy {
             .ignored()
             .or(escape)
             .repeated()
-            .slice()
+            .to_slice()
             .delimited_by(just(b'"'), just(b'"'))
             .boxed();
 
-        let ident = text::ascii::ident().map_slice(Token::Ident);
+        let ident = text::ascii::ident().to_slice().map(Token::Ident);
 
         choice((
             just(b"null").to(Token::Null),

--- a/examples/foo.rs
+++ b/examples/foo.rs
@@ -27,6 +27,7 @@ enum Expr<'a> {
     },
 }
 
+#[allow(clippy::let_and_return)]
 fn parser<'a>() -> impl Parser<'a, &'a str, Expr<'a>> {
     let ident = text::ascii::ident().padded();
 

--- a/examples/json.rs
+++ b/examples/json.rs
@@ -21,12 +21,12 @@ fn parser<'a>() -> impl Parser<'a, &'a str, Json, extra::Err<Rich<'a, char>>> {
     recursive(|value| {
         let digits = text::digits(10).to_slice();
 
-        let frac = just('.').then(digits.clone());
+        let frac = just('.').then(digits);
 
         let exp = just('e')
             .or(just('E'))
             .then(one_of("+-").or_not())
-            .then(digits.clone());
+            .then(digits);
 
         let number = just('-')
             .or_not()

--- a/examples/nano_rust.rs
+++ b/examples/nano_rust.rs
@@ -454,7 +454,7 @@ fn funcs_parser<'tokens, 'src: 'tokens>() -> impl Parser<
             for ((name, name_span), f) in fs {
                 if funcs.insert(name, f).is_some() {
                     emitter.emit(Rich::custom(
-                        name_span.clone(),
+                        name_span,
                         format!("Function '{}' already exists", name),
                     ));
                 }
@@ -489,7 +489,7 @@ fn eval_expr<'src>(
             .map(|(_, v)| v.clone())
             .or_else(|| Some(Value::Func(name)).filter(|_| funcs.contains_key(name)))
             .ok_or_else(|| Error {
-                span: expr.1.clone(),
+                span: expr.1,
                 msg: format!("No such variable '{}' in scope", name),
             })?,
         Expr::Let(local, val, body) => {
@@ -504,20 +504,16 @@ fn eval_expr<'src>(
             eval_expr(b, funcs, stack)?
         }
         Expr::Binary(a, BinaryOp::Add, b) => Value::Num(
-            eval_expr(a, funcs, stack)?.num(a.1.clone())?
-                + eval_expr(b, funcs, stack)?.num(b.1.clone())?,
+            eval_expr(a, funcs, stack)?.num(a.1)? + eval_expr(b, funcs, stack)?.num(b.1)?,
         ),
         Expr::Binary(a, BinaryOp::Sub, b) => Value::Num(
-            eval_expr(a, funcs, stack)?.num(a.1.clone())?
-                - eval_expr(b, funcs, stack)?.num(b.1.clone())?,
+            eval_expr(a, funcs, stack)?.num(a.1)? - eval_expr(b, funcs, stack)?.num(b.1)?,
         ),
         Expr::Binary(a, BinaryOp::Mul, b) => Value::Num(
-            eval_expr(a, funcs, stack)?.num(a.1.clone())?
-                * eval_expr(b, funcs, stack)?.num(b.1.clone())?,
+            eval_expr(a, funcs, stack)?.num(a.1)? * eval_expr(b, funcs, stack)?.num(b.1)?,
         ),
         Expr::Binary(a, BinaryOp::Div, b) => Value::Num(
-            eval_expr(a, funcs, stack)?.num(a.1.clone())?
-                / eval_expr(b, funcs, stack)?.num(b.1.clone())?,
+            eval_expr(a, funcs, stack)?.num(a.1)? / eval_expr(b, funcs, stack)?.num(b.1)?,
         ),
         Expr::Binary(a, BinaryOp::Eq, b) => {
             Value::Bool(eval_expr(a, funcs, stack)? == eval_expr(b, funcs, stack)?)
@@ -532,7 +528,7 @@ fn eval_expr<'src>(
                     let f = &funcs[&name];
                     let mut stack = if f.args.len() != args.0.len() {
                         return Err(Error {
-                            span: expr.1.clone(),
+                            span: expr.1,
                             msg: format!("'{}' called with wrong number of arguments (expected {}, found {})", name, f.args.len(), args.0.len()),
                         });
                     } else {
@@ -546,7 +542,7 @@ fn eval_expr<'src>(
                 }
                 f => {
                     return Err(Error {
-                        span: func.1.clone(),
+                        span: func.1,
                         msg: format!("'{:?}' is not callable", f),
                     })
                 }
@@ -559,7 +555,7 @@ fn eval_expr<'src>(
                 Value::Bool(false) => eval_expr(b, funcs, stack)?,
                 c => {
                     return Err(Error {
-                        span: cond.1.clone(),
+                        span: cond.1,
                         msg: format!("Conditions must be booleans, found '{:?}'", c),
                     })
                 }
@@ -588,10 +584,10 @@ fn main() {
 
         if let Some((funcs, file_span)) = ast.filter(|_| errs.len() + parse_errs.len() == 0) {
             if let Some(main) = funcs.get("main") {
-                if main.args.len() != 0 {
+                if !main.args.is_empty() {
                     errs.push(Rich::custom(
                         main.span,
-                        format!("The main function cannot have arguments"),
+                        "The main function cannot have arguments".to_string(),
                     ))
                 } else {
                     match eval_expr(&main.body, &funcs, &mut Vec::new()) {
@@ -602,7 +598,7 @@ fn main() {
             } else {
                 errs.push(Rich::custom(
                     file_span,
-                    format!("Programs need a main function but none was found"),
+                    "Programs need a main function but none was found".to_string(),
                 ));
             }
         }

--- a/examples/nested.rs
+++ b/examples/nested.rs
@@ -9,6 +9,7 @@ enum Token {
     Parens(Vec<Token>),
 }
 
+#[allow(clippy::let_and_return)]
 fn parser<'a>() -> impl Parser<'a, &'a [Token], i64> {
     recursive(|expr| {
         let num = select_ref! { Token::Num(x) => *x };

--- a/examples/nested_spans.rs
+++ b/examples/nested_spans.rs
@@ -11,6 +11,7 @@ enum Token {
 
 type TokenTreeInput<'a> = SpannedInput<Token, SimpleSpan, &'a [(Token, SimpleSpan)]>;
 
+#[allow(clippy::let_and_return)]
 fn parser<'a>() -> impl Parser<'a, TokenTreeInput<'a>, i64> {
     recursive(|expr| {
         let num = select_ref! { Token::Num(x) => *x };

--- a/src/container.rs
+++ b/src/container.rs
@@ -812,6 +812,7 @@ mod test {
         for idx in 0..C::LEN {
             C::write(&mut uninit, idx, idx);
         }
+        // SAFETY: All elements were initialized.
         unsafe { C::take(uninit) }
     }
 
@@ -820,6 +821,7 @@ mod test {
         for idx in 0..(C::LEN / 2) {
             C::write(&mut uninit, idx, idx);
         }
+        // SAFETY: All elements up to this point were initialized.
         unsafe { C::drop_before(&mut uninit, C::LEN / 2) };
     }
 

--- a/src/input.rs
+++ b/src/input.rs
@@ -6,13 +6,15 @@
 //! ways: from strings, slices, arrays, etc.
 
 pub use crate::stream::{BoxedExactSizeStream, BoxedStream, Stream};
-use core::cell::RefCell;
 
 use super::*;
 #[cfg(feature = "memoization")]
 use hashbrown::HashMap;
 #[cfg(feature = "std")]
-use std::io::{BufReader, Read, Seek};
+use std::{
+    cell::RefCell,
+    io::{BufReader, Read, Seek},
+};
 
 /// A trait for types that represents a stream of input tokens. Unlike [`Iterator`], this type
 /// supports backtracking and a few other features required by the crate.

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -3027,7 +3027,7 @@ mod tests {
         }
 
         let mut chars = String::new();
-        for c in parser().parse_iter(&"abcdefg").into_result().unwrap() {
+        for c in parser().parse_iter("abcdefg").into_result().unwrap() {
             chars.push(c);
         }
 
@@ -3409,7 +3409,7 @@ mod tests {
             many_as.into_iter().collect()
         }
 
-        assert_eq!(parser().parse("aaa").into_result().unwrap(), ());
+        assert_eq!(parser().parse("aaa").into_result(), Ok(()));
     }
 
     #[test]
@@ -3437,10 +3437,7 @@ mod tests {
                 let s = "hello".to_string();
 
                 assert_eq!(parser.get().parse(&s).into_result(), Ok("hello"));
-                assert!(matches!(
-                    parser.get().parse("goodbye").into_result(),
-                    Err(_)
-                ));
+                assert!(parser.get().parse("goodbye").into_result().is_err());
             }
         }
 
@@ -3452,10 +3449,7 @@ mod tests {
                 let parser = crate::cache::Cache::new(MyCache);
 
                 assert_eq!(parser.get().parse(&s).into_result(), Ok("hello"));
-                assert!(matches!(
-                    parser.get().parse("goodbye").into_result(),
-                    Err(_)
-                ));
+                assert!(parser.get().parse("goodbye").into_result().is_err());
             }
         }
     }

--- a/src/number.rs
+++ b/src/number.rs
@@ -130,9 +130,9 @@ mod tests {
         #[test]
         fn subnorm() {
             for bits in 0u32..(1 << 21) {
-                let single: f32 = unsafe { core::mem::transmute(bits) };
+                let single: f32 = f32::from_bits(bits);
                 validate(&format!("{:e}", single));
-                let double: f64 = unsafe { core::mem::transmute(bits as u64) };
+                let double: f64 = f64::from_bits(bits as u64);
                 validate(&format!("{:e}", double));
             }
         }

